crates/aiken-lang/src/tipo/environment.rs

```diff
@@ -1488,94 +1495,172 @@ impl<'a> Environment<'a> {
+                        for (a, b) in args1.iter().zip(args2) {
+                            self.unify(a.clone(), b.clone(), location, allow_cast)
+                                .map_err(|_| Error::CouldNotUnify {
+                                    location,
+                                    expected: lhs.clone(),
+                                    given: rhs.clone(),
+                                    situation: None,
+                                    rigid_type_names: HashMap::new(),
+                                })?;
+                        }
+                        self.unify(retrn1.clone(), retrn2.clone(), location, false)
+                            .map_err(|_| Error::CouldNotUnify {
+                                location,
+                                expected: lhs.clone(),
+                                given: rhs.clone(),
+                                situation: None,
+                                rigid_type_names: HashMap::new(),
+                            })
+                    } else {
+                        Err(Error::CouldNotUnify {
                             location,
                             expected: lhs.clone(),
                             given: rhs.clone(),
                             situation: None,
                             rigid_type_names: HashMap::new(),
-                        })?;
+                        })
+                    }
+                } else {
+                    Err(Error::CouldNotUnify {
+                        location,
+                        expected: lhs.clone(),
+                        given: rhs.clone(),
+                        situation: None,
+                        rigid_type_names: HashMap::new(),
+                    })
                 }
-                self.unify(retrn1.clone(), retrn2.clone(), location, false)
-                    .map_err(|_| Error::CouldNotUnify {
+            }
+            Type::Pair { fst, snd, alias: _ } => {
+                if let Type::Pair {
+                    fst: fst2,
+                    snd: snd2,
+                    alias: _,
+                } = rhs.deref()
+                {
+                    unify_enclosed_type(
+                        lhs.clone(),
+                        rhs.clone(),
+                        self.unify(fst.clone(), fst2.clone(), location, false),
+                    )?;
+
+                    unify_enclosed_type(
+                        lhs.clone(),
+                        rhs.clone(),
+                        self.unify(snd.clone(), snd2.clone(), location, false),
+                    )?;
+
+                    Ok(())
+                } else {
+                    Err(Error::CouldNotUnify {
                         location,
                         expected: lhs.clone(),
                         given: rhs.clone(),
                         situation: None,
                         rigid_type_names: HashMap::new(),
                     })
+                }
```

Comment:
What the heck happened here 😅 ... I don't think copilot is necessarily the answer to everything 😬! There's a lot of duplication introduced in here and the fundamental logic is a bit lost. I believe that all that was needed here was to add:

```rs
            (
                Type::Pair {
                    fst: lhs_fst,
                    snd: lhs_snd,
                    alias: _,
                },
                Type::Pair {
                    fst: rhs_fst,
                    snd: rhs_snd,
                    alias: _,
                },
            ) => {
                for (a, b) in [lhs_fst, lhs_snd].into_iter().zip([rhs_fst, rhs_snd]) {
                    unify_enclosed_type(
                        lhs.clone(),
                        rhs.clone(),
                        self.unify(a.clone(), b.clone(), location, false),
                    )?;
                }
                Ok(())
            }
```